### PR TITLE
Simplify JSON forwarding proposal

### PR DESCRIPTION
**Pull request by alanconway**
- Move `structured`, `schema` fields to pipeline, simplifies use.
- Added `schemaKey` for simpler & more flexible schema by namespace or label.

### CODE REVIEW

**Review by jcantrill — COMMENTED** [review]
/approve

**Comment by alanconway** [comment]
Fixed the lint problems, @jcantrill  LGTM please.

File: enhancements/cluster-logging/forwarding-json-structured-logs.md [hunk 1]
```diff
@@ -123,132 +115,128 @@ Two new top-level fields added to the record [data model][data_model]:
...
+#### ClusterLogForwarder configuration
 
-`input.stuctured`: (string, default "JSON") Attempt to parse structured entries.
-The only legal value is "JSON", others may be added in future.
+New *pipeline* fields:
 
-An input record contains:
+* `stuctured`: (string, optional) Legal values are "JSON" or "json" (there may be others in future).
+   If set, attempt to parse log entries as JSON objects.
+* `schemaKey`: (string, optional) Use meta-data key as `schema` value if present.
```
Comment (sichvoge) [inline enhancements/cluster-logging/forwarding-json-structured-logs.md:133]: This sounds much more complicated :) @jcantrill what are your thoughts?

File: enhancements/cluster-logging/forwarding-json-structured-logs.md [hunk 2]
```diff
@@ -123,132 +115,128 @@ Two new top-level fields added to the record [data model][data_model]:
...
+  - `kubernetes.namespaceName`: Use the namespace name as the schema name.
```
Comment (sichvoge) [inline enhancements/cluster-logging/forwarding-json-structured-logs.md:134]: @alanconway what are the use cases for introducing the `schemaKey` field? not sure i understand correctly to which of the scenarios this one specifically applies.

Specifically, how does this correlate with the `namespace` key for a pipeline to only consider logs from a particular namespace?

Comment (alanconway) [reply enhancements/cluster-logging/forwarding-json-structured-logs.md:134]: The input selectors (log type, namespace, eventually pod labels) are statically configured to apply a user-chosen schema name to an arbitrary set of containers. If you want to add a new schema, you have to change the forwarder config. If you have a lot of labels, your config will be very big.
The schemaKey dynamically applies a metadata value as the schema. You can define a *single* input with schemaKey=namespace and the namespace will be used as the schema. No need to change the CRD every time someone adds a namespace.
IMO the most likely and cleanest use of this whole setup is to create a single CRD with `schemaKey=kubernetes.label.MySchemaLabel`. That means you can just apply MySchemaLabel=whatever to your pods to have whatever labels you want at run-time, no need to mess with the CRD when new schema are defined.
If we don't have schemaKey, then we absolutely need  label selectors, based on the AMDocs/Avaloq discussoins there's really no hope that namespaces will be sufficient. schemaKey is much easier to implement and easier to use.

File: enhancements/cluster-logging/forwarding-json-structured-logs.md [hunk 3]
```diff
@@ -34,52 +34,46 @@ see-also:
...
 
-To illustrate, given a log entry `{"name":"fred","home":"bedrock"}` from a container with the label `app.kubernetes.io/name="flintstones"`. The current output record is:
+To illustrate, given a log entry `{"name":"fred","home":"bedrock"}` from a container with the label `app.kubernetes.io/name="flintstones"`.
+The current output record looks like:
 
 ```json
 {
   "message":"{\"name\":\"fred\",\"home\":\"bedrock\"}",
-  "kubernetes":{"labels":["app_kubernetes_io/name=flintstones", ...}},
+  "kubernetes":{"labels":["app_kubernetes_io/name=flintstones", ...]},
```
[the remark this answers is not in the document]
Comment (alanconway) [reply enhancements/cluster-logging/forwarding-json-structured-logs.md:50]: +1 - but do we have to suppress the map version when forwarding to elasticsearch? @ewolinetz ?

File: enhancements/cluster-logging/forwarding-json-structured-logs.md [hunk 4]
```diff
@@ -123,132 +115,128 @@ Two new top-level fields added to the record [data model][data_model]:
...
-With an *Elasticsearch* store, `schema` is copied to the existing `viaq_index_name` field to set the index.
-`viaq_index_name` will continue to be used for compatibility and upgrade, but will eventually be removed.
-
-#### ClusterLogForwarder configuration
+* The presence of `structured` or `message` fields identifies structured/unstructured records.
+* The `schema` field is optional.
+  It is provided to represent the schema of a structured record, but its use is up to the consumer.
 
-The elements of the CLF API are:
+For the default *Elasticsearch* store, `schema` is used to generate the index name.
```
[the remark this answers is not in the document]
Comment (alanconway) [reply enhancements/cluster-logging/forwarding-json-structured-logs.md:122]: Thats the idea - schema represents a stream of consistent JSON, so we want to send it to the same index, and not to indexes expecting unstructured or other schema JSON.

File: enhancements/cluster-logging/forwarding-json-structured-logs.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (jcantrill) [reply enhancements/cluster-logging/forwarding-json-structured-logs.md:50]: I was commenting here is this is not what we do currently?  we use "flat_labeles"

[the remark this answers is not in the document]
Comment (jcantrill) [reply enhancements/cluster-logging/forwarding-json-structured-logs.md:50]: we should suppress until will add proper support in managed ES.  I'm not certain if we should flatten for unmanaged ES

**Review by portante — CHANGES_REQUESTED** [review]
Very laudable goal, but this proposal seems to do this as part of the flow for logging.  That adds an element of risk as the places were logs are collected are where they are transformed.

Instead, it might be worth looking at doing this outside of the log flow itself as a separate process.

To that end, the logging system would need to accurately capture all the data and metadata so that the log record could be re-processed after in lands the final store.

This would allow the log collection system to focus on what it needs to do well: collect the logs and forward them.

The another component can handle the reprocessing.

This new component could have:

 * N sub-components that read from a supported log store (Elasticsearch, Loki, Splunk, CloudWatch, etc.)
 * a single common sub-component the transforms the read data
 * N sub-components that write to a supported log store

This way, if the transformation needs to be corrected or enhanced, it can be done without affecting the core flow of logs.

**Comment by sichvoge** [comment]
@portante I think that's pretty much related to some investigations @jcantrill is driving around what we call "central normalization". He already finished a small PoC (https://issues.redhat.com/browse/LOG-703) you might be interested to talk about with him. 

The log forwarding API will then move out from the collector (as i think you said) to that component. And since we are talking about an API, we can apply this to the current architecture and change it easily afterwards.

**Comment by sichvoge** [comment]
LGTM

**Comment by jcantrill** [comment]
> Instead, it might be worth looking at doing this outside of the log flow itself as a separate process.

This is tangential to this proposal and fits into central normalization.  @sichvoge has a todo (I think) to assemble this into a more coherent epic/story/initative

> To that end, the logging system would need to accurately capture all the data and metadata so that the log record could be re-processed after in lands the final store.

We can do that now but still are limited to what is provided by the kublet which is: namespace, podname, container hash

**Comment by alanconway** [comment]
> Very laudable goal, but this proposal seems to do this as part of the flow for logging. That adds an element of risk as the places were logs are collected are where they are transformed.
> 
> Instead, it might be worth looking at doing this outside of the log flow itself as a separate process.

As things stand now we have to do this in the logging flow (specifically we have to do it in fluentd) because it changes the format of what fluentd forwards. It's an urgent customer need.
 
> To that end, the logging system would need to accurately capture all the data and metadata so that the log record could be re-processed after in lands the final store.
> This would allow the log collection system to focus on what it needs to do well: collect the logs and forward them.
> 
> The another component can handle the reprocessing.
> 
> This new component could have:
> 
>     * N sub-components that read from a supported log store (Elasticsearch, Loki, Splunk, CloudWatch, etc.)
> 
>     * a single common sub-component the transforms the read data
> 
>     * N sub-components that write to a supported log store
> 
> 
> This way, if the transformation needs to be corrected or enhanced, it can be done without affecting the core flow of logs.

+1, that's where we want to go: a  simpler collector that just forwards bytes, and separate normalizer & forwarder that know how to route logs and what (if any) transformation needs to be done.

**Comment by portante** [comment]
> It's an urgent customer need.

This seems to be a decision to work on the urgent vs the essential.  This runs the risk of perpetuating the issues in the logging stack, which can lead to further delays to implementing the essential building blocks necessary for a healthy logging stack.

If we focus our time on getting an efficient collector to an efficient store, and then we can build visuals indexing logs on demand into Elasticsearch, or a database, or some other visual tool to make it all scale.

And we also solve the policy / rate problem, we can focus on the fun stuff of working with metadata and building observability tools that really help.

**Pull request merged**